### PR TITLE
Last implementation of package classification was very slow. It needed

### PR DESCRIPTION
**Pull request by guliashvili**
linear time in terms of number of different possible packages. My
implementation does preprocessing. It creates array and uses key as
index. However each array can be leaf or can need another array. (Some
packaged need 3 key or 2 key)

### CODE REVIEW

**Comment by guliashvili** [comment]
Hi Guys, As you see it's my first commit here. It's pretty nice project I'd love to optimize many stuff you help me of course :)

**Comment by guliashvili** [comment]
anyone? knock knock :( :-1:

**Comment by Zealar** [comment]
Hello can post some test results? Also please use l2j formatter for code.

**Comment by guliashvili** [comment]
Hi,
What kind of test results?
Where can I find  l2j formatter for code?

**Comment by Zealar** [comment]
You say "Last implementation of package classification was very slow." so what is real difference? 1ms 2ms? What we win if change it to that way.
Is inside https://github.com/L2J/L2J_Server/tree/develop/.settings
is auto format code when use eclipse.

**Comment by guliashvili** [comment]
There is about 500 package, finding random package using last implementation needed checking 250 package on average. With my implementation it's done with cost of 2-5 package check.

File: src/main/java/com/l2jserver/gameserver/network/fastforwarder/FastForwarderTable.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import com.l2jserver.gameserver.network.L2GameClient;
+import com.l2jserver.mmocore.ReceivablePacket;
+import com.l2jserver.util.Util;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.logging.Logger;
+
+public abstract class FastForwarderTable implements FastForwarder
+{
+	private static final Logger _log = Logger.getLogger(FastForwarderTable.class.getName());
```
Comment (Zoey76) [inline src/main/java/com/l2jserver/gameserver/network/fastforwarder/FastForwarderTable.java]: Logger should use SLF4J.

**Comment by Zoey76** [comment]
Do you want to use this to replace which code exactly?

L2GamePacketHandler#handlePacket(ByteBuffer buf, L2GameClient client) ?

**Comment by guliashvili** [comment]
@Zoey76 
Before, whole checking of what packet we got was in 
	[public ReceivablePacket<L2GameClient> handlePacket(ByteBuffer buf, L2GameClient client)](https://github.com/L2J/L2J_Server/blob/develop/src/main/java/com/l2jserver/gameserver/network/L2GamePacketHandler.java)

Now,
I do linear preprocessing and build static table, table is kind of hashmap, but because keys are small I can even do faster thing and use poor arrays without hashes and use keys directly as an array indexes.
So, new [handlePacket function](https://github.com/guliashvili/L2J_Server/blob/develop/src/main/java/com/l2jserver/gameserver/network/L2GamePacketHandler.java#L649) is just one line. Which uses this pre created forwarding table.
Basically I changed everything in L2GamePacketHandler and also added some helper classes.

**Comment by Zoey76** [comment]
If you want to micro-optimize the packet selection, you should read, if you haven't already, about the under-the-hood implementations of switch statement in Java.

[3.10. Compiling Switches ](http://docs.oracle.com/javase/specs/jvms/se7/html/jvms-3.html)

Java uses two modes tableswitch, with Big-O of O(1), corresponding to your implementation and lookupswitch which uses a sorted pair table, with Big-O of O(log(n)), which for 500 packets would be 2.6~ top operations.

How do you plan to create the List<FastForwarderPair> list ?

**Comment by guliashvili** [comment]
Good job java optimizations. :) .
I will check how exactly it is optimized and if my optimization is faster

**Comment by Sacrifice78** [comment]
Any news or improvement about that or any comment at respect???

**Comment by UnAfraid** [comment]
That wouldn't increase performance a lot unless there are 2000-3000 players, even then you might not notice much of a difference
